### PR TITLE
2022_10_31

### PR DESCRIPTION
**Pull request by Kukuru-Dza**

### CODE REVIEW

File: first_lessons_dushanbe/archive/fibonacci_numbers.py [hunk 1]
```diff
@@ -2,20 +2,33 @@
 # -*- coding: utf-8 -*-
 
 def get_fibonacci_number(a):
+    res = []
     if a <= 1:
+        return res.append(0)
+    if a == 2:
+        res = [0, 1]
+        return res
+    res = [0, 1]
```
Comment (shuvava) [inline first_lessons_dushanbe/archive/fibonacci_numbers.py:11]: you does not need to cache all numbers only the last two

File: first_lessons_dushanbe/recursion/recursive_string.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+def rec_str(a):
+    for i in range(0, (len(a) - 1) // 2):
+        a[i] = a[i].replace(a[i], a[i] + '(')
```
Comment (shuvava) [inline first_lessons_dushanbe/recursion/recursive_string.py:6]: `a[i]` is getting access to the char of the string by char index. and you replace this char by line from two char. it means that length of the string will be changed. this logic looks very suspicious

File: first_lessons_dushanbe/recursion/recursive_task_2.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# -*- coding: utf-8 -*-
+
+def rew(a):
+    if len(a) == 1 and a[0] != '(':
+        return a
+    elif len(a) == 1 and a[0] == '(':
+        return ')'
+    else:
+        if a[-1] == '(':
+            return ')' + rew(a[:-1])
```
Comment (shuvava) [inline first_lessons_dushanbe/recursion/recursive_task_2.py:11]: `a[:-1]` is actually make a copy of your string. It means that you using O(n**2) space. This logic can be done in O(1) space.

File: first_lessons_dushanbe/recursion/recursive_string.py (hunk 2 above)
Comment (Kukuru-Dza) [reply first_lessons_dushanbe/recursion/recursive_string.py:6]: я ж ей не строку передаю а список, так что все норм. я просто тренировался в списковых методах.

File: first_lessons_dushanbe/recursion/recursive_task_2.py (hunk 3 above)
Comment (Kukuru-Dza) [reply first_lessons_dushanbe/recursion/recursive_task_2.py:11]: а как?

File: first_lessons_dushanbe/tasks_from_lessons/tabl.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+'''просто тренировочная таблица умножения на генераторы списков'''
+
+if __name__ == '__main__':
+    a = [[0] * 10 for i in range(10)]
+    for i in range(10):
+        for j in range(10):
+            a[i][j] = (i + 1) * (j + 1)
+
+    for i in range(10):
+        print(a[i])
```
Comment (shuvava) [inline first_lessons_dushanbe/tasks_from_lessons/tabl.py]: print has a syntax to add leading zeros if required. Can you modify the output   to straight columns?

**Review by shuvava — APPROVED** [review]

File: first_lessons_dushanbe/tasks_from_lessons/tabl.py (hunk 4 above)
Comment (Kukuru-Dza) [reply first_lessons_dushanbe/tasks_from_lessons/tabl.py]: done

**Pull request merged**